### PR TITLE
[rllib] Fix atari reward calculations, add LR annealing, explained var stat for A2C / impala

### PR DESCRIPTION
**Pull request by ericl**
## What do these changes do?

Changes needed to reproduce Atari plots in IMPALA / A2C: https://github.com/ray-project/rl-experiments

## Related issue number

https://github.com/ray-project/ray/issues/2663

### CODE REVIEW

File: python/ray/rllib/models/lstm.py [hunk 1]
```diff
@@ -162,16 +156,13 @@ def _build_layers(self, inputs, num_outputs, options):
...
-            state_in = rnn.LSTMStateTuple(c_in, h_in)
-        else:
-            state_in = rnn.rnn_cell.LSTMStateTuple(c_in, h_in)
         lstm_out, lstm_state = tf.nn.dynamic_rnn(
             lstm,
             last_layer,
-            initial_state=state_in,
             sequence_length=self.seq_lens,
-            time_major=False)
+            time_major=False,
```
Comment (ericl) [inline python/ray/rllib/models/lstm.py]: @richardliaw I believe initial_state was set incorrectly to state_in before?

File: python/ray/rllib/env/atari_wrappers.py [hunk 2]
```diff
@@ -10,6 +10,68 @@ def is_atari(env):
...
+    while True:
+        if isinstance(currentenv, cls):
+            return currentenv
+        elif isinstance(currentenv, gym.Wrapper):
+            currentenv = currentenv.env
+        else:
+            return None
+
+
+class MonitorEnv(gym.Wrapper):
```
Comment (richardliaw) [inline python/ray/rllib/env/atari_wrappers.py]: I think it would be reasonable to note that this is not the same as gym.Wrapper.Monitor

**Review by richardliaw — APPROVED** [review]
some nits for documentation

File: python/ray/rllib/agents/ppo/ppo_policy_graph.py [hunk 3]
```diff
@@ -179,9 +188,11 @@ def __init__(self,
             self.kl_coeff,
             entropy_coeff=self.config["entropy_coeff"],
             clip_param=self.config["clip_param"],
-            vf_loss_coeff=self.config["kl_target"],
+            vf_loss_coeff=self.config["vf_loss_coeff"],
```
Comment (richardliaw) [inline python/ray/rllib/agents/ppo/ppo_policy_graph.py]: nice

File: python/ray/rllib/agents/a3c/a3c.py [hunk 4]
```diff
@@ -83,8 +85,7 @@ def _init(self):
                 A3CTorchPolicyGraph
             policy_cls = A3CTorchPolicyGraph
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/agents/a3c/a3c.py]: Not really, the variable is static and is only used so users can override it. Updating it is too late once the class is created.

File: python/ray/rllib/env/atari_wrappers.py (hunk 2 above)
Comment (ericl) [reply python/ray/rllib/env/atari_wrappers.py]: Yes

File: python/ray/rllib/evaluation/sampler.py [hunk 5]
```diff
@@ -384,6 +391,24 @@ def new_episode():
         async_vector_env.send_actions(dict(actions_to_send))
 
 
+def _fetch_atari_metrics(async_vector_env):
+    """Atari games have multiple logical episodes, one per life.
```
[the remark this answers is not in the document]
Comment (ericl) [reply python/ray/rllib/evaluation/sampler.py]: You need to distinguish between env types right?

File: python/ray/rllib/agents/a3c/a3c.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/rllib/agents/a3c/a3c.py]: What variable are you referring to as static?

The main point here is that
```
agent = a3c.A3CAgent(use_pytorch=True)
print(agent._policy_graph) 
# prints A3CPolicyGraph
```
is odd.
